pkg/controllermanager/controller/plant/plant_control.go

```diff
@@ -157,6 +157,7 @@ func (r *plantReconciler) reconcile(ctx context.Context, log logr.Logger, plant
 		if err := controllerutils.StrategicMergePatchAddFinalizers(ctx, gardenClient, plant, FinalizerName); err != nil {
 			return fmt.Errorf("failed to ensure finalizer on plant: %w", err)
 		}
+		return nil
```

Comment:
I saw that for the plant we do the following check before adding it if it is updated: https://github.com/gardener/gardener/blob/710ca095fa8c88fd9bf3589f2879beb236e2610d/pkg/controllermanager/controller/plant/plant_control.go#L98-L100

But in the update strategy the generation is only increased if the `plant.spec` is changed: https://github.com/gardener/gardener/blob/710ca095fa8c88fd9bf3589f2879beb236e2610d/pkg/registry/core/plant/strategy.go#L49-L71

If I am not wrong, this means that after the generation is updated for the first reconciliation, adding a finalizer here will not update it again and therefore the `Plant` will not be readded to the queue.

Btw for most of the other resources we do the following comparison in the update method
`new.Generation == new.Status.ObservedGeneration`

Reply:
Thanks @plkokanov, exactly such findings were also my main concern (though, I haven't spent the time yet to check each and every controller). @shafeeqes please consider this and test each controller accordingly so that we don't introduce regressions with this change.

Reply:
Yes. I will check and make changes accordingly.